### PR TITLE
Add device lifecycle proposal

### PR DESCRIPTION
**Pull request by zolkis**
Based on comments in #476

### CODE REVIEW

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+- possible previous states + stakeholders + conditions that controls the state transition
+- possible next states + stakeholders+conditions that control the state transition.
+
+
+### Manufactured state
+**Why**: device is manufactured and flashed with a SW image. Possibly certified (e.g. OCF).
+**Name in other protocols** : manufactured
+**Actor** responsible to get to this state: manufacturer
+**Capabilities**:
+- manufacturer certificate: OCF: optional, OneM2M: optional, LwM2M: optional, T2TRG: optional, OPC-UA: optional, Anima: required
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:61]: thanks! changed

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+- identity: usually none, possibly manufacturer-default
+- operational level: only for onboarding (OCF), initial provisioning (M2M), bootstrap (LwM2M), bootstrapping (installation, commisioning in T2TRG, enrollment in OPC-UA, bootstrapping in Anima)
+- external interfaces: native mechanisms (e.g. support for onboarding methods/discovery)
+
+**Previous state**: none
+**Preceding states** (from which one can reset to Manufactured state): any state except Destroyed
+**Next states**: Bootstrapped/Onboarded
+**Relationship** (why to change state):
+  - onboard the device in order to be usable by a provider (or to re-badge by vendor)
+  - security setup
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:71]: I just remove it indeed. Security is part of the previous relationship.

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+- operational level: native, usually modified SW image, run after rebooting the device. After onboarding is complete, it can identify and authenticate, can be configured for services.
+- operational for WoT: no
+- external interfaces: native mechanisms (e.g. support for further provisioning/configuration)
+
+**Previous state**: Manufactured
+**Preceding states** (from which one can reset to Bootstrapped state): Operational/Maintenance
+**Next states**: Operational, Manufactured
+**Relationship** (why to change state to Operational):
+  - provision the device in order to be usable by a provider in a solution
+  - security setup
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:91]: removed

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+**Actor** who triggers next state transition: provider/vendor, based on native built-in mechanisms and protocols.
+
+### Onboarded state (for given protocol)
+**Why**: device gets owned, gets identity and authorization capability.
+This state may be skipped from WoT point of vew.
+**Name in other protocols** onboarding (OCF), initial provisioning (M2M), bootstrapping (LwM2M), bootstrapping (installation, commisioning) in T2TRG, bootstrapping (enrollment, trust list acquisition) in OPC-UA), bootstrapping in Anima
+**Actor** responsible to get to this state: provider/vendor
+**Capabilities**:
+- identity: given by new owner (provider/vendor or user)
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:81]: changed to provider/vendor/user credentials

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  - "offboarding" the device, i.e. reset to factory defaults
+
+**Actor** who triggers next state transition: provider/vendor or user, based on native built-in mechanisms and protocols.
+
+### Operational state (for given protocol)
+**Why**: device gets operational in its native protocol (provisioned and configured).
+** Name in other protocols ** : operational
+**Actor** responsible to get to this state: provider/vendor
+**Capabilities**:
+- identity: given by new owner (provider/vendor or user)
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:103]: changed

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+  - "offboarding" the device, i.e. reset to factory defaults
+
+**Actor** who triggers next state transition: provider/vendor or user, based on native built-in mechanisms and protocols.
+
+
+### Operational state for WoT
+**Why**: device gets operational for WoT
+**Actor** responsible to get to this state: provider/vendor or user
+**Capabilities**:
+- identity: given by provider/vendor or user
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:130]: changed

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+- provisioned for the solution
+- configured for the solution
+- operational level: native, fully functional.
+- operational for WoT: yes (including deploying WoT runtime, dependencies and system configuration, optionally WoT bootstrapping for WoT services, e.g. Thing Directory)
+- external interfaces: mainly WoT interface
+
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:136]: included

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+**Previous state**: Operational for [protocol]
+**Preceding states**: Operational for [protocol] , Maintenance (Non-operational for maintenance)
+**Next states**: Maintenance, Operational for [protocol], Onboarded (de-commission), Manufactured (reset to factory defaults)
+**Relationship** (why to change state to Maintenance):
+  - run the native mechanisms for SW update or reconfiguration or re-provisioning that requires stopping the normal operation of the device, including eventual rebooting)
+  - managed either by the underlying native SW stack (to update that), or by WoT runtime (only to update WoT, if that requires stopping WoT services)
+  - security setup
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:143]: removed

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+- what identities the state can be linked to
+- what security data the state can be linked to
+- capabilities of the state, e.g.:
+    - is the state operational for the "native" protocol
+    - is the state operational for WoT (note that if the device can be _configured_ to work with WoT without changing the state in the "native" protocol, then it is operational in WoT, too).
+- possible previous states + stakeholders + conditions that controls the state transition
+- possible next states + stakeholders+conditions that control the state transition.
+
+
+### Manufactured state
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:56]: data may be associated to each stakeholder I guess

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 10]
```diff
@@ -1,72 +1,75 @@
...
 
-### Stakeholders
+### Candidate stakeholder names
 
-Also, main stakeholders need to be defined. Since these are inherited from native protocols that try to solve the same use cases as the other protocols, we can establish a common terminology:
+The main stakeholders are inherited from native protocols, which in turn are trying to solve the same use cases as the other protocols and therefore use stakeholder and state names that enable establishing a common terminology:
 - _manufacturer_
 - "vendor" or "service provider" or "site" or "solution provider", perhaps _provider_ being the best
-- _application_
+- _application_ (from the native protocols context)
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:29]: WoT is an application for OCF, for instance. I could mention that as an example.

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 11]
```diff
@@ -1,72 +1,75 @@
...
-
+3. WoT should be able to _describe the security model_ of the device (in any state) as security is one of the key aspects during IoT device lifecycle. Since different protocols might use different terminology, a mapping of terms used in WoT is needed to be included.
+4. WoT may want to be also able to _describe state transitions_, especially what needs to be done to a device to bring it to operational state in its native protocol domain. An attempt on this is made in this document.
 
-### Stakeholders
+### Candidate stakeholder names
 
-Also, main stakeholders need to be defined. Since these are inherited from native protocols that try to solve the same use cases as the other protocols, we can establish a common terminology:
+The main stakeholders are inherited from native protocols, which in turn are trying to solve the same use cases as the other protocols and therefore use stakeholder and state names that enable establishing a common terminology:
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:26]: Hmm, ok. I think I reversed stakeholders (humans) with actors (can be non-human factors that affect the system). 
http://fserranocs460.blogspot.com/2014/01/use-cases-actors-vs-stakeholders.html

[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:26]: I made another attempt, and added _cloud application_ :)

File: proposals/lifecycle-model-proposal-zolkis.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+### Destroyed state
+**Why**: device gets destroyed with no further possibility of functioning
+**Name in other protocols**
+End-of-life (Anima, OPC-UA)
+T2TRG: there is `removed and replaced` but it's possible to recommission, therefore this state does not exist in T2TRG.
+OneM2M, LwM2M: not defined
+OCF: not explicitly defined
+**Actor** responsible to get to this state: provider/vendor or user, enforcement agent
+**Capabilities**: none
+**Previous state**: any
```
[the remark this answers is not in the document]
Comment (zolkis) [reply proposals/lifecycle-model-proposal-zolkis.md:265]: Included text for that.

**Comment by zolkis** [comment]
Updated the [diagram](https://drive.google.com/file/d/1E7dF65r8i1rx5ykJujMYe14RYPOvXAPf/view) with state transitions.

**Comment by clerley** [comment]
What is the target audience for this document? I am thinking TD stands for "Thing Description" but, I am not sure if that is correct. Are the terms defined elsewhere? Can we provide a link?

i.e. we don't **defines** names for those states in Wot, (There is a typo, it should have been "define" instead of defines.

This reads a little confusing to me: "WoT needs SW update that needs to disrupt the WoT interfaces." (I think what you mean is that some software updates will require changes to the network interface. That will in turn cause disruptions to the WoT interfaces which are layered on top of the network interface, is that correct?)

What state would the WoT be in for change of ownership? Maintenance?

**Comment by zolkis** [comment]
> What is the target audience for this document? 

The .md file is a textual explainer behind the diagram linked above, and is meant for the Architecture task force members as input to the lifecycle discussion (i.e. trying to explain my thinking, based on multiple protocols that have been studied for this).

Thanks for the review, I will sweep for typos and broken sentences. 
WoT interfaces are network interfaces that serve WoT interactions.
The update requires changes that need to shut down the services (i.e. there is a service disruption, which - in my thinking - justifies the separate state Maintenance).

> What state would the WoT be in for change of ownership? Maintenance?

For change of ownership one has to redo onboarding. Maintenance is only for SW or config updates that _disrupt_ the services.

**Comment by mlagally** [comment]
Reviewed in Arch call on 14.5.

**Pull request merged**